### PR TITLE
sync: the `include` and `exclude` options consistent with rsync behavior

### PR DESCRIPTION
**Pull request by zhijian-pro**

### CODE REVIEW

**Comment by zhijian-pro** [comment]
#1542

File: pkg/sync/sync.go [hunk 1]
```diff
@@ -661,51 +662,82 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
+			if rule.include {
+				break
+			} else {
+				logger.Debugf("exclude %s", o.Key())
+				return
+			}
+		}
+	}
+	r <- o
+	logger.Debugf("%s is included", o.Key())
```
Comment (davies) [inline pkg/sync/sync.go:738]: We will see it anyway.

File: pkg/sync/sync.go [hunk 2]
```diff
@@ -668,11 +668,20 @@ type rule struct {
 }
 
 func initRules() (rules []rule) {
-	for idx, arg := range os.Args[:len(os.Args)-1] {
-		if arg == "--include" || arg == "-include" {
+	l := len(os.Args)
+	for idx, arg := range os.Args {
+		if l-1 > idx && arg == "--include" || arg == "-include" {
```
Comment (davies) [inline pkg/sync/sync.go:673]: l-1 > idx && (arg == "--include" || arg == "-include")

File: cmd/sync.go [hunk 3]
```diff
@@ -281,6 +281,9 @@ func isS3PathType(endpoint string) bool {
 
 func doSync(c *cli.Context) error {
 	setup(c, 2)
+	if c.IsSet("include") && !c.IsSet("exclude") {
+		logger.Warnf("The include option needs to be used with the exclude option,so the result of the current sync may not match your expectations")
```
Comment (SandyXSD) [inline cmd/sync.go:285]: xxx, or the result xxx

File: docs/zh_cn/reference/command_reference.md [hunk 4]
```diff
@@ -510,6 +510,10 @@ juicefs sync [command options] SRC DST
 - `BUCKET[.ENDPOINT]`：数据存储服务的访问地址，不同存储类型格式可能不同，具体请参考[文档](how_to_setup_object_storage.md#支持的存储服务)。
 - `[/PREFIX]`：可选，源路径和目标路径的前缀，可用于限定只同步某些路径中的数据。
 
+:::tip
+如果想要在 `SRC` 或者 `DST` 中表达文件夹的概念时，请确保路径是以"/"或者"\"结尾的，否则将会被认为是对象名的前缀。
```
Comment (SandyXSD) [inline docs/zh_cn/reference/command_reference.md:514]: not clear

Comment (SandyXSD) [inline docs/zh_cn/reference/command_reference.md:514]: not clear

File: docs/zh_cn/reference/command_reference.md [hunk 5]
```diff
@@ -549,7 +553,13 @@ juicefs sync [command options] SRC DST
 跳过包含 PATTERN (POSIX正则表达式) 的对象名
 
 `--include PATTERN`<br />
-仅同步包含 PATTERN (POSIX正则表达式) 的对象名
+需要与`--exclude` 配合使用。不排除匹配 PATTERN  (POSIX正则表达式) 的文件
+
+
```
Comment (SandyXSD) [inline docs/zh_cn/reference/command_reference.md:558]: too many empty lines

Comment (SandyXSD) [inline docs/zh_cn/reference/command_reference.md:558]: too many empty lines

File: docs/zh_cn/reference/command_reference.md [hunk 6]
```diff
@@ -549,7 +553,13 @@ juicefs sync [command options] SRC DST
 跳过包含 PATTERN (POSIX正则表达式) 的对象名
 
 `--include PATTERN`<br />
-仅同步包含 PATTERN (POSIX正则表达式) 的对象名
+需要与`--exclude` 配合使用。不排除匹配 PATTERN  (POSIX正则表达式) 的文件
```
Comment (SandyXSD) [inline docs/zh_cn/reference/command_reference.md:556]: PATTERN is not a POSIX regular expressions any more

Comment (SandyXSD) [inline docs/zh_cn/reference/command_reference.md:556]: PATTERN is not a POSIX regular expressions any more

File: pkg/sync/sync.go [hunk 7]
```diff
@@ -661,51 +662,90 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
+		k, p = key, pattern
+	}
+	return
+}
+
+// Consistent with rsync behavior, the matching order is adjusted according to the order of the "include" and "exclude" options
+func matchObject(r chan object.Object, rules []rule, o object.Object) {
+	for _, rule := range rules {
+		p, k := alignPatternAndKey(rule.pattern, o.Key())
+		match, err := filepath.Match(p, k)
```
Comment (SandyXSD) [inline pkg/sync/sync.go:733]: use path.Match

Comment (SandyXSD) [inline pkg/sync/sync.go:733]: use path.Match

File: pkg/sync/sync.go [hunk 8]
```diff
@@ -661,51 +662,90 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
 }
 
-func findAny(s string, ps []*regexp.Regexp) bool {
-	for _, p := range ps {
-		if p.FindString(s) != "" {
-			return true
+func initRules() (rules []rule) {
+	l := len(os.Args)
+	for idx, arg := range os.Args {
+		if l-1 > idx && (arg == "--include" || arg == "-include") {
```
Comment (SandyXSD) [inline pkg/sync/sync.go:673]: no need to check idx since it's checked in main

File: pkg/sync/sync.go [hunk 9]
```diff
@@ -661,51 +662,90 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
+				rules = append(rules, rule{pattern: s[1], include: true})
+			}
+		} else if strings.HasPrefix(arg, "--exclude=") || strings.HasPrefix(arg, "-exclude=") {
+			if s := strings.Split(arg, "="); len(s) == 2 && s[1] != "" {
+				rules = append(rules, rule{pattern: s[1], include: false})
+			}
 		}
 	}
-	return false
+	return
```
Comment (SandyXSD) [inline pkg/sync/sync.go:687]: delete this line

File: pkg/sync/sync.go [hunk 10]
```diff
@@ -661,51 +662,90 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
-			if findAny(o.Key(), exc) {
-				logger.Debugf("exclude %s", o.Key())
-				continue
-			}
-			if len(inc) > 0 && !findAny(o.Key(), inc) {
-				logger.Debugf("%s is not included", o.Key())
-				continue
-			}
-			r <- o
+			matchObject(r, rules, o)
```
Comment (SandyXSD) [inline pkg/sync/sync.go:697]: if includeObject(rules, o) {
    r <- o
}

Comment (SandyXSD) [inline pkg/sync/sync.go:697]: if includeObject(rules, o) {
    r <- o
}

File: cmd/sync.go [hunk 11]
```diff
@@ -46,13 +46,32 @@ func cmdSync() *cli.Command {
 This tool spawns multiple threads to concurrently syncs objects of two data storages.
 SRC and DST should be [NAME://][ACCESS_KEY:SECRET_KEY@]BUCKET[.ENDPOINT][/PREFIX].
 
+Include/exclude pattern rules:
+The include/exclude rules each specify a pattern that is matched against the names of the files that are going to be transferred.  These patterns can take several forms:
+
+- if the pattern ends with a / then it will only match a directory, not a file, link, or device.
+- rsync chooses between doing a simple string match and wildcard matching by checking if the pattern contains one of these three wildcard characters: '*', '?', and '[' .
```
Comment (SandyXSD) [inline cmd/sync.go:53]: adjust

File: pkg/sync/sync.go [hunk 12]
```diff
@@ -661,51 +663,96 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
-				continue
-			}
-			r <- o
 		}
 		close(r)
 	}()
 	return r
 }
 
+func alignPatternAndKey(pattern, key string) (k string) {
```
Comment (SandyXSD) [inline pkg/sync/sync.go:714]: Maybe simplified like this:
```
func alignPatternAndKey(pattern, key string) string {
	sep := "/"
	l := strings.Count(pattern, sep) + 1
	ps := strings.Split(key, sep)
	if len(ps) < l {
		return key
	} else if strings.HasSuffix(pattern, sep) {
		return strings.Join(ps[:l-1], sep) + sep
	} else {
		return strings.Join(ps[:l], sep)
	}
}
```

**Review by davies — APPROVED** [review]

**Pull request merged**